### PR TITLE
Web3signer: Add an HTTP Client Wrapper to Interact with Remote Server

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

Part of #9994

**What does this PR do? Why is it needed?**

This PR is the first in a series that aims to add support for Consensys' [Web3Signer](https://docs.web3signer.consensys.net/en/latest/) remote signing server, supported by various eth2 clients today. Web3signer is an important standard for institutional adoption of staking, and we should support it in Prysm today. 

This first PR introduces a new keymanager type in the Prysm validator known as remote-web3signer, and an HTTP client wrapper used internally by the new keymanager. This defines the necessary methods to perform http calls to the remote signer server according to the API spec they define in their docs [here](https://consensys.github.io/web3signer/web3signer-eth2.html).

**NOTE**: This PR does is scoped only to the http client functionality, and is the first of several remaining PRs

### CODE REVIEW

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:19]: Should we have a shorter timeout? There are 12 seconds per slot and we wouldnt want to take up too much waiting for a hanging API response. I propose 3 seconds

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:34]: Errors are typically lowercase unless they are top-level logs
```
log.Error("Could not do %v", err)
```
```suggestion
		return nil, errors.Wrap(err, "invalid Format, unable to parse url")
```

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:88]: This can probably be its own func as well called unmarshalResponse which takes in something to unmarshal into, such as `target interface{}`

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:154]: is there a type alias around http method to make this a bit more strict? Maybe http.Method is an alias over string we could use

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:173]:
```suggestion
func closeBody(body io.Closer) {
```
io Closer is enough here

**Review by rauljordan — COMMENTED** [review]
Let's remove /tmp/validator that was checked into this PR

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:97]:
```suggestion
	return bls.SignatureFromBytes([]byte(decoded))
```

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:132]:
```suggestion
	if _, err := client.doRequest(http.MethodPost, client.BasePath+requestPath, nil); err != nil {
```

File: validator/keymanager/remote-web3signer/client.go
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:69]: // Sign makes an API request to the remote web3signer server and returns the BLS signature.

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:31]: Rather than adding additional layers of custom types, I would recommend using the standard http library patterns. http.Client is already designed to make this easy. Your test can reach into the client object and replace the `Transport` value with a custom `RoundTripper`: https://pkg.go.dev/net/http#RoundTripper

Which allows you to define a function in your tests that has access to the request value and can return any response value.

File: validator/keymanager/remote-web3signer/client_test.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client_test.go:34]: Implementing these tests with a RoundTripper should be more flexible than replacing a package level variable for each test case. For example if you wanted to expose more of the http.Client methodset, or implement table-driven tests.

**Review by kasey — APPROVED** [review]
Really nice work! Only feedback is to take a look at the RoundTripper interface as mentioned in review comments. Congrats on your first golang PR :)

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:31]: will look int making this change

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:87]: A little late for this comment since I already approved, but I was just working on some http client code myself and it occurred to me that I didn't see any http status code checking in this PR. Don't we need to handle non-200 responses?

**Review by kasey — CHANGES_REQUESTED** [review]

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:87]: yes it will need to be added mirroring the spec. was going to handle in future prs, this one has been happy path.

Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:87]: nvm I can probably handle this before we merge it same with the other comment

Comment (kasey) [reply validator/keymanager/remote-web3signer/client.go:87]: k, lmk if I should switch back to approved, if you want to handle status as a follow-up. I just switched to "changes requested" to make sure you had a chance to respond before it got merged. but probably easy enough to add now and less overhead than another review :man_shrugging:

Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:82]: made changes

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:85]: nitpick: you don't need the `else if` when you early return, the "else" is implicit in the fact that you got past the previous return conditoin. this would be equivalent and imo is easier to visually parse:
```
if err != nil {
	return nil, err
}
if resp.StatusCode == 404 {
        return nil, errors.Wrap(err, "public key not found")
}
if resp.StatusCode == 412 {
	return nil, errors.Wrap(err, "signing operation failed due to slashing protection rules")
}
```

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:104]: check status?

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:131]: check status?

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [inline validator/keymanager/remote-web3signer/client.go:140]: check status?

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [reply validator/keymanager/remote-web3signer/client.go:82]: nice

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:104]: oh these only need the 2 statuses that are checked in doRequest

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:131]: same as previous comment

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:140]: same as previous comment

File: validator/keymanager/remote-web3signer/client.go
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/client.go:85]: ok will update this

File: validator/keymanager/remote-web3signer/client.go
Comment (kasey) [reply validator/keymanager/remote-web3signer/client.go:106]: cool, i missed that :)

**Review by kasey — APPROVED** [review]

**Pull request merged**
